udf/extract_canonical_glean_app_info.sql

```diff
@@ -0,0 +1,71 @@
+/*
+
+Returns canonical, human-understandable identification info for Glean sources.
+
+Currently supports only Fenix variants, and it should be expected that the output
+of this function evolves over time. If we rename products, we may choose to update
+the values here so that analyses consistently get the new name.
+
+The first argument (app_id) can be fairly fuzzy; it is tolerant of actual Google
+Play Store appId values like 'org.mozilla.firefox_beta' (mix of periods and underscores)
+as well as BigQuery dataset names with suffixes like 'org_mozilla_firefox_beta_stable'.
+
+The second argument (app_build_id) should be the value in client_info.app_build.
+
+*/
+CREATE OR REPLACE FUNCTION udf.extract_canonical_glean_app_info(app_id STRING, app_build_id STRING)
+RETURNS STRUCT<app_name STRING, channel STRING, canonical_app_id STRING> AS (
+  CASE
+  -- Note that order is important here; we move from more specific names to more general
+  -- so that we can properly ignore dataset suffixes like _stable or _live or _derived.
+  WHEN
+    app_id LIKE 'org_mozilla_fennec_aurora%'
+  THEN
+    STRUCT('Fenix', 'nightly', 'org.mozilla.fennec_aurora')
+  WHEN
+    app_id LIKE 'org_mozilla_firefox_beta%'
+  THEN
+    STRUCT('Fenix', 'beta', 'org.mozilla.firefox_beta')
+  WHEN
+    app_id LIKE 'org_mozilla_firefox%'
+  THEN
+    STRUCT('Fenix', 'release', 'org.mozilla.firefox')
+  WHEN
+    app_id LIKE 'org_mozilla_fenix_nightly%'
+  THEN
+    STRUCT('Firefox Preview', 'nightly', 'org.mozilla.fenix.nightly')
+  WHEN
+    app_id LIKE 'org_mozilla_fenix%'
+  THEN
+    IF(
+      udf.fenix_build_to_datetime(app_build_id) < '2020-07-03',
+      STRUCT('Firefox Preview', 'beta', 'org.mozilla.fenix'),
+      STRUCT('Fenix', 'nightly', 'org.mozilla.fenix')
+    )
+  ELSE
+    ERROR("Bad data")
```

Comment:
Is there a way to test for this case?

Reply:
I meant to fill this out with a better message. But no, I'm not aware of a way at this point to write a test case that exercises this. If we raise an error, the test fails.